### ISSUE TITLE
Remove top-level expressions

### ISSUE THREAD

**Issue by jameshfisher** [puffnfresh/roy#63]
Roy currently allows expressions, eg. console.log, at the top level of a module. This is lax and un-functional. Expressions should be restricted to inside functions (and at some point hopefully within an IO monad). A single main function should be allowed per module.

**Comment by piranha** [comment]
Departing too far from JS semantics isn't good; making it impossible to execute certain instructions of JS will make Roy more limited in it's capabilities than otherwise.

-1.

**Comment by jameshfisher** [comment]
What things would this make impossible? Even conservative languages like Dart enforce a main function. A single entry point is just good practice.

**Comment by piranha** [comment]
require.js would be impossible without top-level statements, taking this from stuff I work with.

And, in case you can counter this example with that Roy can have it's own module format - having one single true format for modules (which can't be improved without changing the language) will make Roy incompatible with the rest of the world (and more difficult to add to existing projects).

Also, I believe that Dart is bad example. Java has no functions, only methods, should anybody follow this convention?

**Comment by jameshfisher** [comment]
I don't feel the need for a Roy module system, for the same reasons as you; a compiled Roy module should be able to be wrapped in any old module boilerplate. In what I'm proposing, this boilerplate would simply include a call to main(); at the end of the compiled JS module. The main function in the Roy code is where you'd add your exports and whatever else.

**Comment by piranha** [comment]
I don't see how this makes easier to integrate with current JS eco system. How would you implement module for require.js then?

**Comment by lexden** [comment]
Having checked out RequireJS again, I'm not sure. :)  How would you currently implement it?

**Comment by lexden** [comment]
I suppose the following would be a current implementation:

```
let module dep1 dep2 =
  {export1: something, export2: somethingelse}

define ["dep1", "dep2"] module
```

Would the following work as a require.js module?:

```
let module dep1 dep2 =
  {export1: something, export2: somethingelse}

let main =
  define ["dep1", "dep2"] module
```

Which would (in my proposal) compile to the JS:

``` javascript
var module = function(dep1, dep2) {
    return {
        "export1": something,
        "export2": somethingelse
    };
}
define(["dep1", "dep2"], module);
```

**Comment by jameshfisher** [comment]
(Sorry, `lexden` is me. :)

**Comment by jameshfisher** [comment]
To make myself perfectly clear: everything in the (nullary) `main` function would simply be dumped at the bottom of the compiled JS.  Therefore the only thing this proposal prohibits is the use of top-level expressions that precede definitions -- would anyone need these?

**Comment by esehara** [comment]
I accept a single entry point is just great practice, Really. :)

But I rather like the way of Schame or Python.It is because it is easy to write the way of it. I think, when required, the method of realizing a main function is made,but when not required, run with top-level expressions .

**Comment by jameshfisher** [comment]
The issue here is whether or not Roy should encourage JS-style imperative coding, where the programmer thinks in terms of flow of execution ("OK, I defined this function here, but redefine it here, so ...").  I would argue that it should not: it is by design a static functional language, where one does not, and should not have to, think this way.

The stated goal of sticking to Javascript _semantics_ is a separate from the issue of Javascript _style_, I think.

**Comment by piranha** [comment]
> To make myself perfectly clear: everything in the (nullary) main function would simply be dumped at the bottom of the compiled JS.

Ah, in this case, it's not that bad when I think about it.

> How would you currently implement it?

Something like this, of course:

```
define ['dep1', 'dep2'], \dep1 dep2 ->
  {export1: something, export2: somethingelse}
```

It's a bit better than having main in this case because you see all dependencies at the top of the module. But that's quite minor I guess...

**Comment by jameshfisher** [comment]
You should be able to define `main` anywhere, including at the top of your module.  So the following should be equivalent to my earlier suggestion:

```
let main =
  define ["dep1", "dep2"] module

let module dep1 dep2 =
  {export1: something, export2: somethingelse}
```

Which would again be equivalent to

```
let main =
  define ["dep1", "dep2"], \dep1 dep2 →
    {export1: something, export2: somethingelse}
```

Where the former should generate

``` javascript
var module = function(dep1, dep2) {
    return {
        "export1": something,
        "export2": somethingelse
    };
};
define(["dep1", "dep2"], module);
```

and the latter would generate

``` javascript
define(["dep1", "dep2"], function(dep1, dep2) {
    return {
        "export1": something,
        "export2": somethingelse
    };
});
```

which I think is exactly what we want.

**Comment by piranha** [comment]
Indeed, in such case this seems almost good. 'Almost' because if someone decides to call 'main' in some function, defined in the same file, we've got a problem... Not that I think this is normal, but this is valid corner case.

**Comment by jameshfisher** [comment]
If we want this (I'd actually be in favour of it), we can instead compile

```
let main () =
  define ["dep1", "dep2"] module

let module dep1 dep2 =
  {export1: something, export2: somethingelse}
```

to

``` javascript
var main = function() {
  define(["dep1", "dep2"], module);
}
var module = function(dep1, dep2) {
    return {
        "export1": something,
        "export2": somethingelse
    };
};
main();
```

I've modified the Roy `main` to take a value of unit type, which is current valid Roy and generates the desired nullary function.  Therefore my proposal becomes extremely simple:
- Disallow top-level expressions.
- If a function `let main () = ...` is defined in the module, the single line `main();` is appended to the bottom of the compiled JS.

I should note that an optimizer like the Closure compiler should reduce something of the form

``` javascript
var main = function() {
  define([...], module);
}
var module = function(...) {
  // do stuff
}
main();
```

to the more compact

``` javascript
define([...], function(...) {
  // do stuff
});
```

**Comment by jameshfisher** [comment]
I just checked; Closure does do that.  It compiles

``` javascript
var main = function() {
define(["dep1", "dep2"], module);
}
var module = function(dep1, dep2) {
  console.log(dep1["foo"]);
  console.log(dep2["bar"]);
}
main();
```

to

``` javascript
define(["dep1", "dep2"], function(a, b) {
  console.log(a.foo);
  console.log(b.bar)
});
```

So I would call my suggestion of dumping the contents of `main` at the end of the file a premature optimization.  I am now in favour of the simple call to `main()`.
